aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+        let observed_lockup_cycle = pool.observed_lockup_cycle;
+        table::borrow_mut(&mut pool.inactive_shares, observed_lockup_cycle)
+    }
+
+    /// Buy shares into active pool on behalf of delegator `shareholder` who
+    /// delegated `coins_amount` stake or reactivated(redeemed) it from pending-inactive pool.
+    fun buy_in_active_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount == 0) return 0;
+
+        pool_u64::buy_in(&mut pool.active_shares, shareholder, coins_amount)
+    }
+
+    /// Execute the pending withdrawal of `delegator_address` on delegation pool `pool`
+    /// if existing and already inactive to allow the creation of a new one.
+    /// `pending_inactive` stake would be left untouched even if withdrawable and should
+    /// be explicitly withdrawn by delegator
+    fun execute_pending_withdrawal(pool: &mut DelegationPool, delegator_address: address) {
+        let (withdrawal_exists, withdrawal_lockup_cycle) = pending_withdrawal_exists(pool, delegator_address);
+        if (withdrawal_exists && withdrawal_lockup_cycle < pool.observed_lockup_cycle) {
+            withdraw_internal(pool, delegator_address, MAX_U64);
+        }
+    }
+
+    /// Buy shares into pending-inactive pool on behalf of delegator `shareholder` who 
+    /// redeemed `coins_amount` from active pool to schedule it for unlocking.
+    /// If there is a pending withdrawal from a past cycle, fail the operation.
+    fun buy_in_inactive_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount == 0) return 0;
+        // execute the pending withdrawal if existing before creating a new one
+        execute_pending_withdrawal(pool, shareholder);
+
+        // save lockup cycle for new pending withdrawal if no existing previous one
+        let observed_lockup_cycle = pool.observed_lockup_cycle;
+        assert!(*table::borrow_mut_with_default(
+            &mut pool.pending_withdrawals,
+            shareholder,
+            observed_lockup_cycle
+        ) == observed_lockup_cycle,
+            error::invalid_state(EPENDING_WITHDRAWAL_EXISTS)
+        );
+
+        // cannot buy inactive shares, only pending_inactive at current lockup cycle
+        pool_u64::buy_in(pending_inactive_shares_pool(pool), shareholder, coins_amount)
+    }
+
+    /// Convert `coins_amount` of coins to be redeemed from share pool `shares_pool`
+    /// to the exact number of shares to redeem in order to achieve this.
+    fun amount_to_shares_to_redeem(
+        shares_pool: &pool_u64::Pool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount >= pool_u64::balance(shares_pool, shareholder)) {
+            // take into account rounding errors and extract entire shares amount
```

Comment:
When would this case happen?

Reply:
the comment here is misleading, we compare to the shareholder's balance to get the maximum shares it can redeem instead of letting the operation fail on:
`assert!(shares(pool, shareholder) >= shares_to_redeem, error::invalid_argument(EINSUFFICIENT_SHARES));`